File: coprocess/grpc/ruby/sample_server.rb
require '../../bindings/ruby/dispatcher'

require '../../bindings/ruby/coprocess_session_state_pb'

require 'json'

class SampleServer < Coprocess::Dispatcher::Service

  # Implements a dynamic dispatcher for CP objects, this class should provide methods for your hooks (see MyPreMiddleware).
  def dispatch(coprocess_object, _unused_call)
    begin
      if self.respond_to?(coprocess_object.hook_name)
        coprocess_object = self.send(coprocess_object.hook_name, coprocess_object)
      else
        raise Coprocess::Dispatcher::HookNotImplemented
      end
    rescue Coprocess::Dispatcher::HookNotImplemented
      puts "Hook not implemented: #{coprocess_object.hook_name}"
    rescue Exception => e
      puts "Couldn't dispatch: #{e}"
    end

    return coprocess_object
  end

  # Implements an event dispatcher.
  def dispatch_event(event_wrapper, _unused_call)
    event = JSON.parse(event_wrapper.payload)
    puts "dispatch_event: #{event}"
    return Coprocess::EventReply.new
  end

  def MyPreMiddleware(coprocess_object)
    coprocess_object.request.set_headers["rubyheader"] = "rubyvalue"
    return coprocess_object
  end

  def MyAuthCheck(coprocess_object)
    puts "Calling MyAuthCheck"
    valid_token = 'aaf4c61ddcc5e8a2dabede0f3b482cd9aea9434d'
    request_token = coprocess_object.request.headers["Authorization"]

    if request_token == nil
      puts "(Using form)"
      request_token = coprocess_object.request.params["key"]
    end

    if request_token.include?(valid_token)
      new_session = Coprocess::SessionState.new
      new_session.rate = 1000
      new_session.per = 10
      new_session.quota_max = 60
      new_session.quota_renews = 1474057703
      new_session.quota_remaining = 0
      new_session.quota_renewal_rate = 120
      new_session.expires = 1474057703

      # Only set this on create!
      new_session.last_updated = (Time.now.to_i + 10).to_s

      # ID Extractor Deadline!
      new_session.id_extractor_deadline = 20

      # new_session.session_lifetime = 99

      coprocess_object.metadata["token"] = "mytoken"
      coprocess_object.session = new_session
    else
      coprocess_object.request.return_overrides.response_code = 401
      coprocess_object.request.return_overrides.response_error = 'Not authorized (gRPC/Ruby middleware)'
    end

    return coprocess_object
  end
end

def main
  s = GRPC::RpcServer.new
  s.add_http2_port('0.0.0.0:5555', :this_port_is_insecure)
  s.handle(SampleServer)
  s.run_till_terminated
end

main

File: coprocess/bindings/python/sample_server.py
import coprocess_object_pb2, coprocess_object_pb2_grpc
from coprocess_session_state_pb2 import SessionState

import grpc, time, json, gzip

_ONE_DAY_IN_SECONDS = 60 * 60 * 24

from concurrent import futures

def MyPreMiddleware(coprocess_object):
  coprocess_object.request.set_headers["MyPreMiddleware"] = "MyPreMiddleware"
  return coprocess_object

def MyPostMiddleware(coprocess_object):
  coprocess_object.response.headers["MyPostMiddleware"] = "MyPostMiddleware"
  return coprocess_object

def MyResponseMiddleware(coprocess_object):
    # Unzip Response
    gzip_decoded_body = gzip.decompress(coprocess_object.response.raw_body)
    json_body = json.loads(str(gzip_decoded_body, 'utf-8'))
    ids = list()

    # Extract IDs
    for item in json_body:
        ids.append(item['id'])

    # Format Response
    coprocess_object.response.body = json.dumps(ids)
    coprocess_object.response.raw_body = json.dumps(ids).encode('utf-8')
    coprocess_object.response.headers['Content-Length'] = str(len(coprocess_object.response.raw_body))
    coprocess_object.response.headers['Content-Encoding'] = ''
    return coprocess_object

def MyAuthCheck(coprocess_object):
  valid_token = 'aaf4c61ddcc5e8a2dabede0f3b482cd9aea9434d'
  request_token = coprocess_object.request.headers["Authorization"]

  if request_token == valid_token:
      new_session = SessionState()
      new_session.rate = 1000.0
      new_session.per = 1.0
      coprocess_object.metadata["token"] = "mytoken"
      coprocess_object.session.CopyFrom(new_session)

  else:
      coprocess_object.request.return_overrides.response_code = 401
      coprocess_object.request.return_overrides.response_error = 'Not authorized (Python middleware)'

  return coprocess_object

class MyDispatcher(coprocess_object_pb2_grpc.DispatcherServicer):
  def Dispatch(self, coprocess_object, context):
    if coprocess_object.hook_name == "MyPreMiddleware":
        coprocess_object = MyPreMiddleware(coprocess_object)

    if coprocess_object.hook_name == "MyPostMiddleware":
        coprocess_object = MyPostMiddleware(coprocess_object)

    if coprocess_object.hook_name == "MyAuthCheck":
        coprocess_object = MyAuthCheck(coprocess_object)

    if coprocess_object.hook_name == "MyResponseMiddleware":
        coprocess_object = MyResponseMiddleware(coprocess_object)

    return coprocess_object

  def DispatchEvent(self, event_wrapper, context):
    event = json.loads(event_wrapper.payload)
    return coprocess_object_pb2_grpc.EventReply()


def serve():
  server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
  coprocess_object_pb2_grpc.add_DispatcherServicer_to_server(
      MyDispatcher(), server)
  server.add_insecure_port('[::]:5555')
  server.start()
  try:
    while True:
      time.sleep(_ONE_DAY_IN_SECONDS)
  except KeyboardInterrupt:
    server.stop(0)

if __name__ == '__main__':
  serve()